cpu/stm32/periph/usbdev_fs.c

```diff
@@ -0,0 +1,665 @@
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    _global_regs(conf)->CNTR = USB_CNTR_CTRM | USB_CNTR_RESETM  |
+                               USB_CNTR_ERRM | USB_CNTR_PMAOVRM;
+}
+
+static void _setup(stm32_usbdev_fs_t *usbdev,
+                      const stm32_usbdev_fs_config_t *config)
+{
+    usbdev->usbdev.driver = &driver;
+    usbdev->config = config;
+    usbdev->out = _ep_out;
+    usbdev->in = _ep_in;
+}
+
+static uint32_t _type_to_reg(usb_ep_type_t type)
+{
+    switch (type) {
+        case USB_EP_TYPE_CONTROL:
+            return USB_EP_CONTROL;
+        case USB_EP_TYPE_ISOCHRONOUS:
+            return USB_EP_ISOCHRONOUS;
+        case USB_EP_TYPE_BULK:
+            return USB_EP_BULK;
+        case USB_EP_TYPE_INTERRUPT:
+            return USB_EP_INTERRUPT;
+        default:
+            assert(false);
+            return 0;
+    }
+}
+
+void usbdev_init_lowlevel(void)
+{
+    for (size_t i = 0; i < USBDEV_NUMOF; i++) {
+        _setup(&_usbdevs[i], &stm32_usbdev_fs_config[i]);
+    }
+}
+
+/*TODO: rework this function (maybe move it outside this file)
+        for other ST families support */
+static void _enable_usb_clk(void)
+{
+#if defined(PWR_CR2_USV)
+    /* Validate USB Supply */
+    PWR->CR2 |= PWR_CR2_USV;
+#endif
+
+#if defined(RCC_APB1SMENR_USBSMEN)
+    RCC->APB1SMENR |= RCC_APB1SMENR_USBSMEN;
+#elif defined(RCC_APB1SMENR_USBSMEN)
+    RCC->APB1SMENR1 |= RCC_APB1SMENR1_USBSMEN;
+#endif
+    /* Enable CRS with auto trim enabled */
+    CRS->CR |=  (CRS_CR_AUTOTRIMEN |  CRS_CR_CEN);
+}
+
+static void _enable_gpio(const stm32_usbdev_fs_config_t *conf)
+{
+    gpio_init(conf->dp, GPIO_IN);
+    gpio_init(conf->dm, GPIO_IN);
+    /* Configure AF for the pins */
+    gpio_init_af(conf->dp, conf->af);
+    gpio_init_af(conf->dm, conf->af);
+}
+
+static void _set_ep_in_status(uint16_t *val, uint16_t mask)
+{
+    /* status endpoints bits can only be toggled, writing 0
+       as no effect. Thus, check which bits should be set
+       then XOR it with their current value */
+    if (mask & USB_EPTX_DTOG1) {
+        *val ^= USB_EPTX_DTOG1;
+    }
+    if (mask & USB_EPTX_DTOG2) {
+        *val ^= USB_EPTX_DTOG2;
+    }
+}
+
+static void _set_ep_out_status(uint16_t *val, uint16_t mask)
+{
+       /* status endpoints bits can only be toggled, writing 0
+       as no effect. Thus, check which bits should be set
+       then XOR it with their current value */
+    if (mask & USB_EPRX_DTOG1) {
+        *val ^= USB_EPRX_DTOG1;
+    }
+    if (mask & USB_EPRX_DTOG2) {
+        *val ^= USB_EPRX_DTOG2;
+    }
+}
+
+static inline void _usb_attach(stm32_usbdev_fs_t *usbdev)
+{
+    /* Some ST USB IP doesn't have this feature */
+#ifdef USB_BCDR_DPPU
+    const stm32_usbdev_fs_config_t *conf = usbdev->config;
+    bool syncing = false;
+    /* Enable DP pullup to signal connection */
+    _global_regs(conf)->BCDR |= USB_BCDR_DPPU;
+    do {
+        syncing = CRS->ISR & 0x00000008;
+    } while (!syncing);
```

Comment:
```suggestion
    while (!(CRS->ISR & CRS_ISR_ESYNCF));
```

?

Reply:
Changed, doesn't seem to break anything.